src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -0,0 +1,945 @@
+                                                const int* Ui){
+
+    int row;
+    for(int i = 0;i<n;++i) {
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i]; j<Lp[i+1]; ++j) {
+        row = Li[j];
+        //BUT dont count dia_gonal twice, important
+        if(row != i) {
+          mia_[row + 1]++;
+        }
+      }
+      //then each column of U
+      for(int j = Up[i];j<Up[i+1];++j) {
+        row = Ui[j];
+        mia_[row + 1]++;
+      }
+    }
+    //then organize mia_;
+    mia_[0] = 0;
+    for(int i=1;i<n+1;i++) {
+      mia_[i] += mia_[i-1];
+    } 
+
+    int* Mshifts = (int*) calloc (n,sizeof(int));
+    for(int i = 0;i<n;++i) {
+
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i];j<Lp[i+1];++j) {
+        row = Li[j];
+        if(row != i) {
+          //place (row, i) where it belongs!
+
+          mja_[mia_[row] + Mshifts[row]] = i;
+          Mshifts[row]++;
+        }
+      }
+      //each column of U next
+      for(int j = Up[i];j<Up[i+1];++j) {
+        row = Ui[j];
+        mja_[mia_[row] + Mshifts[row]] = i;
+        Mshifts[row]++;
+      }
+    }
+    free(Mshifts);
+    return 0;
+  }
+
+
+  // call if both the matrix and the nnz structure changed or if convergence is poor while using refactorization.
+  int hiopLinSolverIndefSparseCUSOLVER::newKLUfactorization()
+  {
+    klu_free_symbolic(&Symbolic_, &Common_) ;
+    klu_free_numeric(&Numeric_, &Common_) ;
+    Symbolic_ = klu_analyze(n_, kRowPtr_, jCol_, &Common_) ;
+    if(Symbolic_ == NULL) {
+      nlp_->log->printf(hovError, "symbolic NULL\n");
+      return -1;
+    }
+
+    Numeric_ = klu_factor(kRowPtr_, jCol_, kVal_, Symbolic_, &Common_);
+    if(Numeric_ == NULL) {
+      nlp_->log->printf(hovError, "printf matrix size: %d numeric NULL \n", n_);
+      return -1;
+    }
+    // get sizes
+    const int nnzL = Numeric_->lnz;
+    const int nnzU = Numeric_->unz;
+
+    const int nnzM = (nnzL + nnzU - n_);
+    /* parse the factorization */
+
+    if(mia_ != NULL) {
+      free(mia_);
+    }
+    if(mja_ != NULL) {
+      free(mja_);
+    }
+
+    mia_ = (int*) calloc(sizeof(int), (n_+1));
+    mja_ = (int*) calloc(sizeof(int), nnzM);
+    int* Lp = (int*) malloc((n_+1) * sizeof(int));
+    int* Li = (int*) malloc(nnzL * sizeof(int));
+    //we cant use NULL instrad od Lx and Ux because it causes SEG FAULT. It seems like a waste of memory though.      
+    double* Lx = (double*) malloc(nnzL * sizeof(double));
+    int* Up = (int*) malloc((n_+1) * sizeof(int));
+    int* Ui = (int*) malloc(nnzU * sizeof(int));
+    double* Ux = (double*) malloc(nnzU * sizeof(double));
+
+    int ok = klu_extract(Numeric_, Symbolic_, Lp, Li, Lx, Up, Ui, Ux, 
+                         NULL, NULL, NULL, 
+                         NULL, NULL, 
+                         NULL, NULL, &Common_);
+
+    createM(n_, nnzL, Lp, Li,nnzU, Up, Ui);
+    free(Lp); 
+    free(Li); 
+    free(Lx);
+    free(Up); 
+    free(Ui); 
+    free(Ux);
```

Comment:
Although IMO we should _really_ be using `std::vector` and `vec.data()` or `std::unique_ptr<T[]>` and `ptr.get()`, I think we should at least be using new/delete over malloc/free.
```suggestion
    mja_ = (int*) calloc(sizeof(int), nnzM);
    int* Lp = new int[n_+1];
    int* Li = new int[nnzL];
    //we cant use NULL instrad od Lx and Ux because it causes SEG FAULT. It seems like a waste of memory though.
    double* Lx = new double[nnzL];
    int* Up = new int[n_+1];
    int* Ui = new int[nnzU];
    double* Ux = new double[nnzU];
    int ok = klu_extract(Numeric_, Symbolic_, Lp, Li, Lx, Up, Ui, Ux, 
                         NULL, NULL, NULL, 
                         NULL, NULL, 
                         NULL, NULL, &Common_);

    createM(n_, nnzL, Lp, Li,nnzU, Up, Ui);
    delete[] Lp;
    delete[] Li;
    delete[] Lx;
    delete[] Up;
    delete[] Ui;
    delete[] Ux;
```